[CALCITE-3139] Implement JSON_EXTRACT function

XuQianJin-Stars opened the pull request:
`JSON_EXTRACT(json_doc, path[, path] ...)`

Returns data from a JSON document, selected from the parts of the document matched by the `path` arguments. Returns `NULL` if any argument is `NULL` or no paths locate a value in the document. An error occurs if the `json_doc` argument is not a valid JSON document or any `path` argument is not a valid path expression.

The return value consists of all values matched by the `path` arguments. If it is possible that those arguments could return multiple values, the matched values are autowrapped as an array, in the order corresponding to the paths that produced them. Otherwise, the return value is the single matched value.

Example Sql:
```
SELECT JSON_EXTRACT(v, '$') AS c1
,JSON_EXTRACT(v, '$.b') AS c2
,JSON_EXTRACT(v, '$.a[0]') AS c3
,JSON_EXTRACT(v, '$.a[0]','$.b[0]') AS c4
FROM (VALUES ('{"a": [10, true],"b": [11]}')) AS t(v)
limit 10
```
Result:

|c1|c2|c3|c4|
| --- | --- | --- | --- |
|{"a": [10, true],"b": "[10, true]"}|11|10|[10,11]|

XuQianJin-Stars committed:
[CALCITE-3139] Implement JSON_EXTRACT function

chunweilei commented on core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java:
```diff
@@ -644,6 +644,35 @@ public static Integer jsonStorageSize(JsonValueContext input) {
+      for (String pathSpec : pathSpecs) {
+        value = ctx.read(pathSpec);
+        if ((pathSpec != null) && (value != null)) {
+          arrayList.add(value);
```
It seems parentheses is not necessary in` (pathSpec != null) && (value != null)`.

XuQianJin-Stars replied:
well,let me change it.

chunweilei commented on core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:
```diff
@@ -11070,11 +11070,18 @@ private void checkCustomColumnResolving(String table) {
     checkFails("select ^json_remove('{\"foo\":\"bar\"}')^",
-            "(?s).*Invalid number of arguments.*");
+        "(?s).*Invalid number of arguments.*");
   }
```
Is this change expected?

XuQianJin-Stars replied:
yes,This is the expected change.

XuQianJin-Stars commented:
hi @chunweilei  Thank you very much, Comments addressed.

jinxing64 commented on core/src/main/java/org/apache/calcite/sql/fun/SqlJsonExtractFunction.java:
```diff
@@ -0,0 +1,73 @@
+
+  @Override public boolean checkOperandTypes(SqlCallBinding callBinding, boolean throwOnFailure) {
+    final int operandCount = callBinding.getOperandCount();
+    assert operandCount >= 2;
```
How about we do this check by overriding `checkOperandCount` ?

XuQianJin-Stars replied:
> How about we do this check by overriding `checkOperandCount` ?

Because this function must at least guarantee two parameters.

XuQianJin-Stars replied:
> How about we do this check by overriding `checkOperandCount` ?

Thank you very much, I can also use what you said. Let me adjust it.

danny0405 committed:
[CALCITE-3331] Support implicit type cast for operators that use single operand family checker

* Add doc to SqlSingleOperandTypeChecker#checkSingleOperandType to note
that we should not support implicit type coercion for it's
implementation;
* Tweak the operands checking logic for SqlJsonRemoveFunction,
SqlRegexpReplaceFunction and SqlSubstringFunction to support implicit
type coercion, also add the test cases in SqlValidatorTest;
* Some cosmetic comments fix.

XuQianJin-Stars committed:
[CALCITE-3139] Implement JSON_EXTRACT function
